beacon-chain/slasher/rpc.go

```diff
@@ -0,0 +1,64 @@
+package slasher
+
+import (
+	"context"
+
+	slashertypes "github.com/prysmaticlabs/prysm/beacon-chain/slasher/types"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// IsSlashableBlock checks if an input block header is slashable
+// with respect to historical block proposal data.
+func (s *Service) IsSlashableBlock(
+	ctx context.Context, block *ethpb.SignedBeaconBlockHeader,
+) (*ethpb.ProposerSlashing, error) {
+	dataRoot, err := block.Header.HashTreeRoot()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get block header hash tree root: %v", err)
+	}
+	signedBlockWrapper := &slashertypes.SignedBlockHeaderWrapper{
+		SignedBeaconBlockHeader: block,
+		SigningRoot:             dataRoot,
+	}
+	proposerSlashings, err := s.detectProposerSlashings(ctx, []*slashertypes.SignedBlockHeaderWrapper{signedBlockWrapper})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not check if proposal is slashable: %v", err)
+	}
+	if len(proposerSlashings) == 0 {
+		return nil, nil
+	}
+	return proposerSlashings[0], nil
+}
+
+// IsSlashableAttestation checks if an input indexed attestation is slashable
+// with respect to historical attestation data.
+func (s *Service) IsSlashableAttestation(
+	ctx context.Context, attestation *ethpb.IndexedAttestation,
+) ([]*ethpb.AttesterSlashing, error) {
+	dataRoot, err := attestation.Data.HashTreeRoot()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get attestation data hash tree root: %v", err)
+	}
+	indexedAttWrapper := &slashertypes.IndexedAttestationWrapper{
+		IndexedAttestation: attestation,
+		SigningRoot:        dataRoot,
+	}
+
+	attesterSlashings, err := s.checkSlashableAttestations(ctx, []*slashertypes.IndexedAttestationWrapper{indexedAttWrapper})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not check if attestation is slashable: %v", err)
+	}
+	if len(attesterSlashings) == 0 {
+		// If the incoming attestations are not slashable, we mark them as saved in
+		// slasher's DB storage to help us with future detection.
+		if err := s.serviceCfg.Database.SaveAttestationRecordsForValidators(
```

Comment:
Saving to DB seems out of scope base on method name and comment. Can the caller do it?

Reply:
We would need to have two RPC endpoints. One for checking an att is slashable and another for committing. I agree that seems like a more sensible option, but for the sake of time, perhaps we can proceed with this and improve once we have an rc?

Reply:
Let's rename the function to reflect it is saving to db?

Reply:
I think that saving to the db, if we have never seen this attestation before, is reasonable. 

However, the reason we original had this into two methods is that we didn't want the slasher to broadcast our own slashable events.